firedrake/mg/opencascade_mh.py

```diff
@@ -0,0 +1,250 @@
+                prolong(Ts[i-1], Ts[i])
+    return mh
+
+
+def make_coarse_mesh(stepfile, cad, element_size, dim, comm=COMM_WORLD, distribution_parameters=None, cache=True, verbose=True, gmsh="gmsh"):
+
+    curdir = os.path.dirname(stepfile) or os.getcwd()
+    stepname = os.path.basename(os.path.splitext(stepfile)[0])
+    geopath = os.path.join(curdir, "coarse-%s.geo" % (stepname))
+    mshpath = geopath.replace("geo", "msh")
+
+    if not os.path.isfile(mshpath) or not cache:
+
+        if comm.rank == 0:
+            geostr = 'SetFactory("OpenCASCADE");\n'
+            geostr += 'a() = ShapeFromFile("%s");\n' % os.path.abspath(stepfile)
+            if isinstance(element_size, tuple):
+                assert len(element_size) == 2
+
+                geostr += """
+Mesh.CharacteristicLengthMin = %s;
+Mesh.CharacteristicLengthMax = %s;
+                """ % (element_size[0], element_size[1])
+            elif isinstance(element_size, int) or isinstance(element_size, float):
+                geostr += """
+Mesh.CharacteristicLengthMin = %s;
+Mesh.CharacteristicLengthMax = %s;
+                """ % (element_size, element_size)
+            elif isinstance(element_size, str):
+                geostr += element_size
+            else:
+                raise NotImplementedError("element_size has to be a tuple, a number or a string")
+
+            if dim == 2:
+                for i in range(1, cad.number_of_edges()+1):
+                    geostr += "Physical Line(%d) = {%d};\n" % (i, i)
+                for i in range(1, cad.number_of_faces()+1):
+                    geostr += ('Physical Surface(%d) = {%d};\n' % (i+cad.number_of_edges(), i))
+                if cad.number_of_faces() > 1:
+                    surfs = "".join(["Surface{%d}; " % i for i in range(2, cad.number_of_faces()+1)])
+                    geostr += ('BooleanUnion{ Surface{1}; Delete;}{' + surfs + 'Delete;}')
+            elif dim == 3:
+                for i in range(1, cad.number_of_faces()+1):
+                    geostr += "Physical Surface(%d) = {%d};\n" % (i, i)
+
+                geostr += ('Physical Volume("Combined volume", %d) = {a()};\n' % (cad.number_of_faces()+1))
+
+            logging.debug(geostr)
+
+            with open(geopath, "w") as f:
+                f.write(geostr)
+
+            try:
+                os.remove(mshpath)
+            except OSError:
+                pass
+
+            if verbose:
+                stdout = None
+            else:
+                stdout = subprocess.DEVNULL
+
+            gmsh = subprocess.Popen(gmsh.split(" ") + ["-%d" % dim, geopath], stdout=stdout)
+            gmsh.wait()
+
+        comm.barrier()
+
+    coarse = Mesh(mshpath, distribution_parameters=distribution_parameters, comm=comm)
+    return coarse
+
+
+def push_coordinates_to_plex(mesh):
+    plex = mesh._plex
+    x = mesh.coordinates
+    with x.dat.vec_ro as x_:
+        plex.setCoordinates(x_)
+        plex.setCoordinatesLocal(x_)
```

Comment:
Two things here:

1. `x_` is a global vector, so you only need to do `plex.setCoordinates` (it has some invalidation mechanism to ensure that the coordinatesLocal are reconstructed later) I think

2. The data layout is, I think, not the same for plex's view of the coordinates and ours (they have differently numbered sections (AKA cell_node_map)), so I don't know how this works!

Reply:
You're right, I can delete the call to `setCoordinatesLocal` and the tests still pass. I've committed this.

Reply:
Do you have any insight into point 2?

Reply:
(I mean, clearly it does work, but I wonder if it does so by accident, magic, or something else).

Reply:
Alas, I'm afraid I don't. @florianwechsung, do you have any insight?

Reply:
Mh no I have no idea. In fact, are we sure it DTRT? Where are the coordinates in plex used apart from when calling refine (which happens before this)? All the assembly and transfers use the coordinates specified in `mesh.coordinates`, right?

Reply:
The setup is something like:

1. Plex has topology + geometry (via DMGetCoordinateDM)
2. When we build a `Mesh` we share (effectively) the topology and build a new geometry, by creating a VFS and _copying_ the coordinates from the plex into the VFS (via a permutation map)
3. Now we ignore the plex coordinates.

When plex refines, it uses the coordinates it knows about the its geometry. These are values (in a Vec) plus data layout (in a section).

When this code presently does `push_coords_to_plex`, it updates the values (with what Firedrake knows), but _not_ the data layout. I think if we want to replace the data layout too, we need to do:

```python
sec = new_coords.function_space().dm.getDefaultSection()
plex_sec = PETSc.Section().create(comm=new_coords.comm)
plex_sec.setNumFields(1)
plex_sec.setFieldComponents(gdim)
plex_sec.setPermutation(mesh._plex_renumbering)
for p in range(*sec.getChart()):
    dof = sec.getDof(p)
    plex_sec.setDof(p, dof*gdim)
    plex_sec.setFieldDof(p, 0, dof*gdim)
plex_sec.setUp()
plex.getCoordinateDM().setDefaultSection(plex_sec)
```
plex.setCoordinatesLocal(

Reply:
untested, not even sure if all of those petsc4py functions exist.

Reply:
I've deleted the `push_coordinates_to_plex` code, as discussed.